src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -166,6 +179,221 @@ instance (NFData t, NFData r) => NFData (SnapMergingRunState t r) where
+        mergeState <- newMVar $! MT.OngoingTreeMerge ongoingState
+        newRef (MT.finalise mergeState) $ \mergeRefCounter ->
+          MT.MergingTree { mergeState, mergeRefCounter }
+
+    getSnapMergingRunState
+      :: forall t.
+         MR.IsMergeType t
+      => NumRuns
+      -> NumEntries
+      -> Int
+      -> SnapMergingRunState t (Ref (Run m h))
+      -> m (Ref (MR.MergingRun t m h))
+    getSnapMergingRunState nr ne sc smrs = do
+          let ln = LevelNo 0
+              caching = diskCachePolicyForLevel confDiskCachePolicy ln
+              alloc = bloomFilterAllocForLevel conf ln
+              indexType = indexTypeForRun confFencePointerIndex
+          case smrs of
+            SnapCompletedMerge run ->
+              withRollback reg (MR.newCompleted nr ne run) releaseRef
+            SnapOngoingMerge runs mt -> do
+              rn <- uniqueToRunNumber <$> incrUniqCounter uc
+              mr <- withRollback reg
+                (MR.new hfs hbio resolve caching alloc indexType mt (mkPath rn) runs)
+                releaseRef
+              -- When a snapshot is created, merge progress is lost, so we
+              -- have to redo merging work here. SuppliedCredits tracks how
+              -- many credits were supplied before the snapshot was taken.
+              leftoverCredits <- MR.supplyCredits
+                                   mr (creditThresholdForLevel conf ln)
+                                   (MR.Credits sc)
+              assert (leftoverCredits == 0) $ return mr
+
+    fromSnapPreExistingRun :: SnapPreExistingRun (Ref (Run m h)) -> m (MT.PreExistingRun m h)
+    fromSnapPreExistingRun (SnapPreExistingRun run) = MT.PreExistingRun <$> dupRun run
+    fromSnapPreExistingRun (SnapPreExistingMergingRun nr ne (SuppliedCredits sc) smrs) =
+        MT.PreExistingMergingRun <$> getSnapMergingRunState nr ne sc smrs
+
+
+{-------------------------------------------------------------------------------
+  Conversion to merge tree snapshot format
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE toSnapMergingTree :: MT.MergingTree IO h -> IO (SnapMergingTree (Ref (Run IO h))) #-}
+toSnapMergingTree ::
+     (PrimMonad m, MonadMVar m)
+  => MT.MergingTree m h
+  -> m (SnapMergingTree (Ref (Run m h)))
+toSnapMergingTree (MT.MergingTree mStateVar _mCounter) =
+  withMVar mStateVar $ \mState -> SnapMergingTree <$> toSnapMergingTreeState mState
+
+{-# SPECIALISE toSnapMergingTreeState :: MT.MergingTreeState IO h -> IO (SnapMergingTreeState (Ref (Run IO h))) #-}
+toSnapMergingTreeState ::
+     (PrimMonad m, MonadMVar m)
+  => MT.MergingTreeState m h
+  -> m (SnapMergingTreeState (Ref (Run m h)))
+toSnapMergingTreeState (MT.CompletedTreeMerge r) = pure $ SnapCompletedTreeMerge r
+toSnapMergingTreeState (MT.PendingTreeMerge p) = SnapPendingTreeMerge <$> toSnapPendingMerge p
+toSnapMergingTreeState (MT.OngoingTreeMerge mergingRun) = do
```

Comment:
Can this reuse existing code for snapshotting merging runs?

It looks like we now have the same code in three places: `toSnapPreExistingRun`, `toSnapIncomingRun` and `toSnapMergingTreeState`.